### DISCUSSION TITLE
Why is AbortController's cancel() called at the end of every request?

### DISCUSSION THREAD

**Discussion by kieranm** [trpc/trpc#4749] [Q&A] [answered]
For background, we use TRPC for server <> server communication, which is not the primary use case, but it does work really well. This post concerns the TRPC client's HTTP link, and talks about **outgoing** requests to other TRPC services from a client perspective. When we initialise the `HTTPLink`, we pass in an `AbortController` from the `adobe-fetch` library. We use `adobe-fetch` because it has the best HTTP/2 support.

We're seeing sporadic `AbortError`s being thrown sometimes (on the client side) when we try to make requests from one service to another. It's very unclear to me why this is happening at the moment, I'm investigating it.

I noticed something curious. The abort() method on the AbortController (passed into the HTTP link as an argument) seems to be being called at the end of every outgoing request, regardless of whether or not that request was successful.

<img width="555" alt="Screenshot 2023-08-26 at 16 50 19" src="https://github.com/trpc/trpc/assets/232283/579cb6d8-888f-45a7-b50b-6066e01f6a56">

Looking in the code here.
https://github.com/trpc/trpc/blob/main/packages/client/src/links/internals/httpUtils.ts#L200-L202

```
export function httpRequest(
  opts: HTTPRequestOptions,
): PromiseAndCancel<HTTPResult> {
  const ac = opts.AbortController ? new opts.AbortController() : null;
  const meta = {} as HTTPResult['meta'];

  const promise = new Promise<HTTPResult>((resolve, reject) => {
    fetchHTTPResponse(opts, ac)
      .then((_res) => {
        meta.response = _res;
        return _res.json();
      })
      .then((json) => {
        meta.responseJSON = json;
        resolve({
          json: json as TRPCResponse,
          meta,
        });
      })
      .catch((err) => {
        reject(TRPCClientError.from(err, { meta }));
      });
  });
  const cancel = () => {
    ac?.abort();
  };
  return { promise, cancel };
}
```

The `cancel` function here ultimately ends up being given to the observable as a cleanup function [here](https://github.com/trpc/trpc/blob/main/packages/client/src/links/httpLink.ts#L76-L78C13)

I didn't expect this, I'm sorry if I'm missing something, but I would only expect `cancel()` to be invoked when the requester actually wants to cancel a request that is currently in flight, and not every time after a request has completed.

Is this behaviour intentional? What is the rationale behind this approach? I can't be certain this has something to do with the errors we're seeing but it feels like it might.

Thanks!

**Comment by KATT** [comment, marked as answer]
This looks like a bug, feel free to open an issue :)

**Reply by kieranm** [reply]
Have converted to an issue, thanks!
